### DISCUSSION TITLE
I still need to press `R` to refresh the neotree after commit the changes

### DISCUSSION THREAD

**Discussion by LintaoAmons** [nvim-neo-tree/neo-tree.nvim#1253] [Q&A] [answered]
I already set this to be true
```
use_libuv_file_watcher=true
```

But I still need to manually press `R` to refresh the neotree
(See the following gif, last part after I commit the changes, I have to press R to make the color in the file tree disappear)

![show](https://github.com/nvim-neo-tree/neo-tree.nvim/assets/95092244/87a61995-fc0d-419b-82a3-cd156fb26c29)

**Comment by cseickel** [comment]
I'd bet that this problem is caused by the window management library you are using, is it edgy? If you did the commit from the filesystem source or at least while t was open, it would probably work. It would also probably work if you used Neotree commands to toggle between the sources as sidebars.

**Reply by LintaoAmons** [reply]
Okie, I found that if I commit from another new tab, it won't refresh. But if I commit from the same tab, it will refresh.

(Since I open diffview's tab to check the changes and then commit from there, so the neotree didn't refresh)

**Comment by LintaoAmons** [comment]
Hi, is there a api to call to refresh neotree?
So I can call it in my commit function

**Reply by ghost** [reply]
@LintaoAmons 
What is your commit function?
I'm trying to solve your same problem but I'm not sure where i need to put the api call that was mentioned in this discussion?

**Reply by ghost** [reply]
Don't worry I figured it out.
https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/1490

**Comment by LintaoAmons** [comment]
HiHi~ Is it possible to have this api?...

**Comment by miversen33** [comment]
If your goal is to refresh the filesystem, the following _should_ work
```lua
require("neo-tree.sources.filesystem.commands").refresh(require("neo-tree.sources.manager").get_state("filesystem"))
```

**Reply by miversen33** [reply, marked as answer]
To expand on this a bit
```lua
-- 
-- Get the commands module from neo-tree.sources.filesystem. Found here: https://github.com/nvim-neo-tree/neo-tree.nvim/blob/main/lua/neo-tree/sources/filesystem/commands.lua
require("neo-tree.sources.filesystem.commands")
-- Call the refresh function found here: https://github.com/nvim-neo-tree/neo-tree.nvim/blob/2f2d08894bbc679d4d181604c16bb7079f646384/lua/neo-tree/sources/filesystem/commands.lua#L11-L13
.refresh(
    -- Pull in the manager module. Found here: https://github.com/nvim-neo-tree/neo-tree.nvim/blob/2f2d08894bbc679d4d181604c16bb7079f646384/lua/neo-tree/sources/manager.lua
    require("neo-tree.sources.manager")
        -- Fetch the state of the "filesystem" source, feeding it to the filesystem refresh call since most everything in neo-tree
        -- expects to get its state fed to it
        .get_state("filesystem")
)
```

**Reply by LintaoAmons** [reply]
Nice ! Exactly what I want!

**Reply by asvtkn** [reply]
Running this command seems to result in neo-tree opening if it was closed. Is there any way to avoid that?

**Reply by cseickel** [reply]
Instead of refreshing, you can just fire the GIT_EVENT, which is honestly better in every way and will work for all sources:

``` lua
local events = require("neo-tree.events")
events.fire_event(events.GIT_EVENT)
```

**Comment by SimonBrandner** [comment]
I still have trouble with this whenever I commit from the terminal (`ToggleTerm`)... Does anyone have any suggestions?

**Reply by ghost** [reply]
@SimonBrandner 
I did this so that whenever I close Lazygit it updates Neotree.
You can just change the pattern to some like "terminal" (I'm not sure what the name of the terminal is but doing something like `pattern = "*term*"` should work).
```lua
vim.api.nvim_create_autocmd({ "BufLeave" }, {
  -- Try changing *lazygit* to *term*
  pattern = { "*lazygit*" },
  group = vim.api.nvim_create_augroup("neovim_update_tree" {clear = true})
  callback = function()
    require("neo-tree.sources.filesystem.commands").refresh(
      require("neo-tree.sources.manager").get_state("filesystem")
    )
  end,
})
```

**Reply by SimonBrandner** [reply]
Do you think it would be able to detect a command being entered into the terminal?

**Reply by ghost** [reply]
Hm, that's definitely a tough one. I wouldn't know fully at the moment, but I think the closest you could get would be detecting if the user presses Enter whilst in the terminal buffer since there is no way for neovim to interact with commands running in ToggleTerm. 
My other idea was to as an example alias `git add` on your system to also have a command that sends some lua to neovim but I don't even know if that is possible.

**Reply by SimonBrandner** [reply]
It all feels so hacky, I wonder if I would have better luck with another plugin since VS Code can do this just fine...

**Reply by cseickel** [reply]
@SimonBrandner, do you have `use_libuv_file_watcher=true`? It should work automatically if you do.

**Reply by SimonBrandner** [reply]
Yes, I do, it does not help, sadly

**Comment by juanibiapina** [comment]
I have the same issue when commiting with lazygit outside of neovim.

The accepted answer seems to require a "trigger". Is there a way to make it happen automatically? On focus change could be a nice option.
